File: backend/app/schemas/organization.py
from datetime import date

from pydantic import BaseModel, Field


class ReportSpaceRead(BaseModel):
    id: str
    workspace_id: str
    space_type: str
    department_id: str | None = None
    project_team_id: str | None = None
    name: str
    status: str
    report_enabled: bool
    report_mode: str = "weekly"
    ai_enabled: bool
    allowed_data_source_types: list[str]
    template_bindings: dict[str, object] = Field(default_factory=dict)
    member_visibility: str


class ReportSpaceConfigUpdate(BaseModel):
    report_enabled: bool | None = None
    report_mode: str | None = Field(default=None, pattern="^(daily|weekly|daily_weekly)$")
    ai_enabled: bool | None = None
    allowed_data_source_types: list[str] | None = None
    template_bindings: dict[str, object] | None = None
    member_visibility: str | None = Field(default=None, pattern="^(private|department)$")


class DepartmentRead(BaseModel):
    id: str
    workspace_id: str
    parent_id: str | None = None
    name: str
    path: str
    depth: int
    sort_order: int
    status: str
    report_space: ReportSpaceRead | None = None


class DepartmentTreeNode(DepartmentRead):
    children: list["DepartmentTreeNode"] = Field(default_factory=list)


class DepartmentCreate(BaseModel):
    parent_id: str | None = None
    name: str = Field(min_length=1, max_length=120)
    sort_order: int = 0


class DepartmentUpdate(BaseModel):
    parent_id: str | None = None
    name: str | None = Field(default=None, min_length=1, max_length=120)
    sort_order: int | None = None
    status: str | None = None


class DepartmentMemberAdd(BaseModel):
    member_id: str


class SubmissionStatusRead(BaseModel):
    report_space_id: str
    total_members: int
    submitted_members: int
    pending_members: int
    overdue_members: int


class ReportRuleRead(BaseModel):
    id: str
    workspace_id: str
    report_space_id: str
    report_type: str
    enabled: bool
    frequency: str
    interval_value: int | None = None
    week_start_day: int | None = None
    reminder_day: int | None = None
    reminder_time: str | None = None
    due_type: str
    due_day: int | None = None
    due_time: str | None = None
    skip_weekends: bool
    notification_channels: list[str]
    overdue_policy: dict[str, object]
    extra_config: dict[str, object]


class ReportRuleUpsert(BaseModel):
    enabled: bool = True
    frequency: str
    interval_value: int | None = Field(default=None, ge=1)
    week_start_day: int | None = Field(default=None, ge=1, le=7)
    reminder_day: int | None = Field(default=None, ge=1, le=7)
    reminder_time: str | None = None
    due_type: str = "same_day"
    due_day: int | None = Field(default=None, ge=1, le=7)
    due_time: str | None = None
    skip_weekends: bool = False
    notification_channels: list[str] = Field(default_factory=lambda: ["in_app", "feishu"])
    overdue_policy: dict[str, object] = Field(default_factory=dict)
    extra_config: dict[str, object] = Field(default_factory=dict)


class ProjectTeamRead(BaseModel):
    id: str
    workspace_id: str
    name: str
    description: str
    goal: str
    status: str
    start_date: date | None = None
    expected_end_date: date | None = None
    actual_end_date: date | None = None
    report_space: ReportSpaceRead | None = None


class ProjectTeamMemberRead(BaseModel):
    project_team_id: str
    member_id: str
    role: str
    display_name: str
    email: str | None = None
    employee_no: str | None = None
    status: str


class ProjectTeamMemberAdd(BaseModel):
    member_id: str
    role: str = Field(default="project_member", pattern="^(project_admin|project_member)$")


class ProjectTeamMemberUpdate(BaseModel):
    role: str = Field(pattern="^(project_admin|project_member)$")


class ProjectTeamCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str = ""
    goal: str = ""
    start_date: date | None = None
    expected_end_date: date | None = None


class ProjectTeamUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    description: str | None = None
    goal: str | None = None
    status: str | None = None
    start_date: date | None = None
    expected_end_date: date | None = None
    actual_end_date: date | None = None

File: frontend/src/api/organization.ts
import { apiGet, apiPost, apiPut } from './client'
import type { ReportInstance } from './reports'

export interface ReportSpace {
  id: string
  workspace_id: string
  space_type: string
  department_id?: string | null
  project_team_id?: string | null
  name: string
  status: string
  report_enabled: boolean
  report_mode: ReportMode
  ai_enabled: boolean
  allowed_data_source_types: string[]
  template_bindings: Record<string, TemplateBinding>
  member_visibility: string
}

export type ReportMode = 'daily' | 'weekly' | 'daily_weekly'

export interface TemplateBinding {
  template_id?: string
  version_policy?: 'latest_published' | 'fixed_version'
  template_version_id?: string
}

export interface DepartmentTreeNode {
  id: string
  workspace_id: string
  parent_id?: string | null
  name: string
  path: string
  depth: number
  sort_order: number
  status: string
  report_space?: ReportSpace | null
  children: DepartmentTreeNode[]
}

export interface SubmissionStatus {
  report_space_id: string
  total_members: number
  submitted_members: number
  pending_members: number
  overdue_members: number
}

export interface ReportRule {
  id: string
  workspace_id: string
  report_space_id: string
  report_type: string
  enabled: boolean
  frequency: string
  interval_value?: number | null
  week_start_day?: number | null
  reminder_day?: number | null
  reminder_time?: string | null
  due_type: string
  due_day?: number | null
  due_time?: string | null
  skip_weekends: boolean
  notification_channels: string[]
  overdue_policy: Record<string, unknown>
  extra_config: Record<string, unknown>
}

export interface ReportRuleUpsert {
  enabled: boolean
  frequency: string
  interval_value?: number | null
  week_start_day?: number | null
  reminder_day?: number | null
  reminder_time?: string | null
  due_type: string
  due_day?: number | null
  due_time?: string | null
  skip_weekends: boolean
  notification_channels: string[]
  overdue_policy: Record<string, unknown>
  extra_config: Record<string, unknown>
}

export function getDepartmentTree() {
  return apiGet<DepartmentTreeNode[]>('/departments/tree')
}

export function getDepartmentSubmissionStatus(departmentId: string) {
  return apiGet<SubmissionStatus>(`/departments/${departmentId}/submission-status`)
}

export function listReportRules(reportSpaceId: string) {
  return apiGet<ReportRule[]>(`/report-spaces/${reportSpaceId}/rules`)
}

export function updateReportRule(
  reportSpaceId: string,
  reportType: 'daily' | 'weekly',
  payload: ReportRuleUpsert
) {
  return apiPut<ReportRule, ReportRuleUpsert>(
    `/report-spaces/${reportSpaceId}/rules/${reportType}`,
    payload
  )
}

export interface ReportSpaceConfigUpdate {
  report_enabled?: boolean
  report_mode?: ReportMode
  ai_enabled?: boolean
  allowed_data_source_types?: string[]
  template_bindings?: Record<string, TemplateBinding>
  member_visibility?: 'private' | 'department'
}

export function updateReportSpaceConfig(reportSpaceId: string, payload: ReportSpaceConfigUpdate) {
  return apiPut<ReportSpace, ReportSpaceConfigUpdate>(
    `/report-spaces/${reportSpaceId}/config`,
    payload
  )
}

export interface GenerateReportInstancesRequest {
  anchor_date?: string | null
}

export interface GenerateReportInstancesResult {
  created: ReportInstance[]
  existing: ReportInstance[]
}

export function generateReportInstances(
  reportSpaceId: string,
  payload: GenerateReportInstancesRequest = {}
) {
  return apiPost<GenerateReportInstancesResult, GenerateReportInstancesRequest>(
    `/report-spaces/${reportSpaceId}/generate-instances`,
    payload
  )
}
